h2/src/main/org/h2/mvstore/tx/Transaction.java

```diff
@@ -376,14 +445,89 @@ long getLogId() {
         return getLogId(statusAndLogId.get());
     }
 
+    /**
+     * Check whether this transaction is open.
+     */
+    private void checkOpen(int status) {
+        if (status != STATUS_OPEN) {
+            throw DataUtils.newIllegalStateException(
+                    DataUtils.ERROR_TRANSACTION_ILLEGAL_STATE,
+                    "Transaction {0} has status {1}, not open", transactionId, status);
+        }
+    }
+
     /**
      * Check whether this transaction is open or prepared.
      */
     void checkNotClosed() {
         if (getStatus() == STATUS_CLOSED) {
             throw DataUtils.newIllegalStateException(
-                    DataUtils.ERROR_CLOSED, "Transaction is closed");
+                    DataUtils.ERROR_CLOSED, "Transaction {0} is closed", transactionId);
+        }
+    }
+
+    void closeIt() {
+        long lastState = setStatus(STATUS_CLOSED);
+        store.store.deregisterVersionUsage(txCounter);
+        if(hasChanges(lastState) || hasRollback(lastState)) {
+            notifyAllWaitingTransactions();
+        }
+    }
+
+    private synchronized void notifyAllWaitingTransactions() {
+        notifyAll();
+    }
+
+    public boolean waitFor(Transaction toWaitFor) {
+        if (isDeadlocked(toWaitFor)) {
+            StringBuilder details = new StringBuilder(String.format("Transaction %d has been chosen as a deadlock victim. Details:%n", transactionId));
```

Comment:
Usually it's a bad idea to use `StringBuilder` together with the string formatter. However, this is not critical in such code that is not executed in normal conditions.

Also `\n` is used everywhere in H2 sources, it's strange to use platform-specific line break only here.